nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -225,17 +232,109 @@ public static boolean isCompatibleDataType(final Object value, final DataType da
+
+        Optional<DataType> inferredDataTypeOptional = Optional.ofNullable(inferDataType(value, null))
+                .filter(dataType -> !dataType.getFieldType().equals(RecordFieldType.STRING));
+
+        if (value instanceof String) {
+            mostSuitableType = findMostSuitableTypeByStringValue((String) value, types, dataTypeMapper);
+        } else if (inferredDataTypeOptional.isPresent()) {
+            DataType inferredDataType = inferredDataTypeOptional.get();
+
+            Optional<T> inferredTypeOptional = types.stream()
+                    .filter(type -> dataTypeMapper.apply(type).equals(inferredDataType))
+                    .findFirst();
+
+            if (inferredTypeOptional.isPresent()) {
+                mostSuitableType = inferredTypeOptional;
+            } else {
+                Optional<T> widerAvailableTypeOptional = types.stream()
+                        .map(type -> getWiderType(dataTypeMapper.apply(type), inferredDataType).isPresent() ? type : null)
+                        .filter(Objects::nonNull)
+                        .findFirst();
+
+                if (widerAvailableTypeOptional.isPresent()) {
+                    mostSuitableType = widerAvailableTypeOptional;
+                } else {
+                    mostSuitableType = Optional.empty();
+                }
+            }
+        } else {
+            mostSuitableType = Optional.empty();
+        }
+
+        return mostSuitableType;
+    }
+
+    public static <T> Optional<T> findMostSuitableTypeByStringValue(String valueAsString, List<T> types, Function<T, DataType> dataTypeMapper) {
+        Optional<T> mostSuitableType = types.stream()
+                // Sorting based on the RecordFieldType enum ordering looks appropriate here as we want simpler types
+                //  first and the enum's ordering seems to reflect that
+                .sorted((type1, type2) -> {
+                            int comparison;
+
+                            RecordFieldType dataType1 = dataTypeMapper.apply(type1).getFieldType();
+                            RecordFieldType dataType2 = dataTypeMapper.apply(type2).getFieldType();
+
+                            // Moving TIMESTAMP at the front (at least it should precede DATE)
+                            if (dataType1 == RecordFieldType.TIMESTAMP) {
+                                comparison = -1;
+                            } else if (dataType2 == RecordFieldType.TIMESTAMP) {
+                                comparison = 1;
+                            } else {
+                                comparison = dataType1.compareTo(dataType2);
+                            }
+
+                            return comparison;
+                        }
+                )
+                .filter(type -> {
+                    boolean compatible;
+
+                    try {
+                        compatible = isCompatibleDataType(valueAsString, dataTypeMapper.apply(type));
```

Comment:
This can be made much simpler by just simply returning the value. I.e., `return isCompatibleDataType(...);` and then `return false` if an Exception is caught.

Reply:
I really like to follow the "one return only" convention because I find returning a simple local variable at the end of the method actually helps reading/understanding the code in 2 ways:
1. Easy to find _what_ is being returned and during debugging it's a user-friendly experience to just put a break point on that single line if one wants to know what is returned value.
2. One can simply click on the variable that is being returned and have the IDE highlight it's usages across the method, helping a lot understanding the flow of the logic (as most of the time highlighting the variable that is being returned is basically highlighting the main decision making points in the method where the returning variable is being assigned a value).
This is really a personal preference (although I think it is beneficial to most). But if there's a different convention that is usually used I can apply that as well.

Reply:
Simplified this as well.